### PR TITLE
Refactor 1924 split widgets into classes

### PR DESCRIPTION
**Pull request by shaysw**
The essence of this PR is to decouple Widgets from the ever-growing inforgraphics_utils (IU).

To do this I:
- Extracted RequestParams to its own class
- Created a widgets directory/module. In it is the Widget class + extracted all relevant methods from IU and placed them in widget_utils
- Created 2 directories/modules under widgets: sub/urban_widgets. Note that all relevant modules are imported in the __init__.py (see below)
- Extracted all widgets from IU and placed each in its own file in the correct module
- Moved all translations to their relevant widget (and re-compiled)
- The test_accident_count_by_car_type API test is currently failing, this is because I was not able to override the method "get_accidents_stats" which was in IU and has now moved to widget_utils. WIP

Things to note:
- widget_dict is now part of the Widget class, and is initialized/accessed by IU when importing the Widget module. I am not sure this is best practice and would love feedback/suggestions
- In order to add new widgets we now need to create the widget class (in a new file), AND ADD IT to the import line of the relevant __init__ file. for example, if I want to add a new SubUrban widget, I need to add it to widgets/suburban_widgets/__init__.py. The docs need to be updated as well
- I tested these changes but have not gone over all widgets, so any additional testing, especially to registered widgets, is most welcome

### This is a very delicate and dangerous effort, so if any and all souls participating in the CR could please be extra vigilant to any changes which I may have accidently put or to things I might have forgot I would be grateful

### CODE REVIEW

File: anyway/infographics_utils.py [hunk 1]
```diff
@@ -2181,25 +159,21 @@ def get_latest_accident_date(table_obj, filters):
...
     # noinspection PyArgumentList
-    for w in widgets_dict.values():
+    for w in Widget.widgets_dict.values():
         if w.is_relevant(request_params) and w.is_in_cache() == to_cache:
             widget: Widget = w(request_params)
             widgets.append(widget)
             logging.debug(f"name:{widget.name}, class:{get_widget_class_by_name(widget.name)}")
     for w in widgets:
-        w.generate_items()
+        print(w.name)
```
Comment (BarVolunteering) [inline anyway/infographics_utils.py:172]: Why to print here?

File: anyway/infographics_utils.py [hunk 2]
```diff
@@ -2181,25 +159,21 @@ def get_latest_accident_date(table_obj, filters):
...
+        try:
+            w.generate_items()
+        except Exception as e:
+            print(e)
```
Comment (BarVolunteering) [inline anyway/infographics_utils.py:176]: Use log instead of `print`?

**Review by BarVolunteering — COMMENTED** [review]
Nice job, There some comments below.
Notes:
1. I assumed that you didn't change the widgets logic - so i didn't review it.
2. There are some changes that seems unrelated to this PR (Black side effects?) - like 'anyway/backend_constants.py' and others, as this a relatively big code change it might be better to change them in a different PR

File: anyway/infographics_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:172]: crap remainder from debugging, should be removed. good catch

File: anyway/infographics_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:176]: logger is actually not defined in info utils, adding it as well

File: anyway/infographics_utils.py [hunk 3]
```diff
@@ -4,1929 +4,40 @@
...
-            raise ValueError("Urban fields missing")
-        super().__init__(request_params, name)
-
-    @staticmethod
-    def is_relevant(request_params: RequestParams) -> bool:
-        return is_urban(request_params)
-
-
-def get_widget_factories() -> List[Callable[[RequestParams], Widget]]:
+def get_widget_factories() -> List[Callable[[RequestParams], type(Widget)]]:
```
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:34]: It wasn't my idea, I kept getting this error. If you (or anyone) has any idea why this is I would love to learn:

`Type Type cannot be instantiated; use type() instead`

image

File: anyway/infographics_utils.py [hunk 4]
```diff
@@ -4,1929 +4,40 @@
...
-    def __init__(self, request_params: RequestParams):
-        super().__init__(request_params, type(self).name)
-        self.rank = 5
-        self.information = (
-            "Fatal accidents distribution by accident type - head on collisions vs other accidents."
-        )
-
-    def generate_items(self) -> None:
-        self.items = self.get_head_to_head_stat()
+    return list(Widget.widgets_dict.values())
```
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:36]: note that the line (36) is for the function `get_widget_factories`, which returns the exact same value as before this refactor, e.g., I haven't changed anything (from master):
image

File: anyway/infographics_utils.py [hunk 5]
```diff
@@ -4,1929 +4,40 @@
...
-                logging.exception(
-                    f"AccidentTypeVehicleTypeRoadComparisonWidget.localize_items: Exception while translating {item}."
-                )
-        return items
-
-
-def run_query(query: db.session.query) -> Dict:
-    # pylint: disable=no-member
-    return pd.read_sql_query(query.statement, query.session.bind).to_dict(orient="records")
+def get_widget_class_by_name(name: str) -> type(Widget):
```
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:39]: see above

File: anyway/infographics_utils.py [hunk 6]
```diff
@@ -2181,25 +159,21 @@ def get_latest_accident_date(table_obj, filters):
...
-    res = {}
-    for e in entity:
-        label = e.get_label()
-        res[e.value] = {"name": label, "localized_name": _(label)}
-    return res
-
-
 # noinspection PyArgumentList
-def generate_widgets(request_params: RequestParams, to_cache: bool = True) -> List[Widget]:
+def generate_widgets(request_params: RequestParams, to_cache: bool = True) -> List[type(Widget)]:
```
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py:163]: see above

File: anyway/infographics_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elazarg) [reply anyway/infographics_utils.py:39]: You need to use square brackets: `Type[Class] `

**Review by OriHoch — COMMENTED** [review]
the changes look fine except:
1. many unrelated code formatting changes, making it harder to review the relevant changes
2. test code changes - I would prefer to have refactoring without any test changes (or very minimal), which allows to verify that tests remain successful after refactoring. Currently there are many test code changes which is not clear why they are needed

**Comment by BarVolunteering** [comment]
There something weird with your branch - there shouldn't be any conflict with the files:
alembic/versions/30862849651c_add_organization_tables.py
anyway/error_code_and_strings.py
anyway/models.py
anyway/scripts/build_redacted_tables_for_dev.py
anyway/user_functions.py

*and maybe some other.
as to my understanding this PR shouldn't change them

File: anyway/infographics_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shaysw) [reply anyway/infographics_utils.py]: thanks, fixed

**Comment by ziv17** [comment]
Regarding naming conventions
I think module names are lowercase letters with underscore between words. I think having mixed naming conventions is something we want to avoid.

**Comment by ziv17** [comment]
Regarding module directory locations. Here too, I think mixed standard is something we want to avoid. If a base class is also a directory where the modules of its sub-classes reside, then we should do it across our project.

File: anyway/widgets/Widget.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    def serialize(self):
+        if not self.items:
+            self.generate_items()
+        output = {"name": self.name, "data": {}}
+        output["data"]["items"] = self.items
+        if self.text:
+            output["data"]["text"] = self.text
+        if self.meta:
+            output["meta"] = self.meta
```
Comment (elazarg) [inline anyway/widgets/Widget.py:90]: you can simply write `self.meta or {}`

File: anyway/widgets/suburban_widgets/AccidentCountByAccidentYearWidget.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                res,
+                AccidentSeverity.codes(),
+                list(
+                    range(
+                        self.request_params.start_time.year, self.request_params.end_time.year + 1
+                    )
+                ),
+            )
+        except Exception as e:
+            logging.exception(f"failed to add empty keys to {res}", e)
```
Comment (elazarg) [inline anyway/widgets/suburban_widgets/AccidentCountByAccidentYearWidget.py:52]: no need to use `e` here at all. the function logs the current exception.

File: anyway/widgets/suburban_widgets/AccidentCountByCarTypeWidget.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                    "car_type": k,
+                    "percentage_segment": data_by_segment[k],
+                    "percentage_country": national_data[k],
+                }
+            )
+        return out
+
+    @staticmethod
+    def percentage_accidents_by_car_type(involved_by_vehicle_type_data):
+        driver_types = defaultdict(float)
```
Comment (elazarg) [inline anyway/widgets/suburban_widgets/AccidentCountByCarTypeWidget.py:65]: `collection.Counter` is more appropriate. And there's no need to use `float`.

File: anyway/widgets/suburban_widgets/AccidentCountByDriverTypeWidget.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                driver_types[
+                    # pylint: disable=no-member
+                    DriverType.PRIVATE_VEHICLE_DRIVER.get_label()
+                ] += count
+            elif (
+                vehicle_type in VehicleCategory.LIGHT_ELECTRIC.get_codes()
+                or vehicle_type in VehicleCategory.OTHER.get_codes()
+            ):
+                # pylint: disable=no-member
+                driver_types[DriverType.OTHER_DRIVER.get_label()] += count
```
Comment (elazarg) [inline anyway/widgets/suburban_widgets/AccidentCountByDriverTypeWidget.py:58]: here and in other places, simply getting the key, and only the updating, would give a cleaner code.

**Comment by elazarg** [comment]
I agree with @ziv17 regarding the naming conventions. Modules should be lowercase, and giving each class a dedicated module is unconventional and unnecessary.

**Comment by OriHoch** [comment]
> I agree with @ziv17 regarding the naming conventions. Modules should be lowercase, and giving each class a dedicated module is unconventional and unnecessary.

I think for this type of cases it's easier to work on each class in a separate file rather then a single big file
I don't know of any convention regarding it, I know it's quite common, for example in many places of the Django code such a convention is used - https://github.com/django/django/tree/main/django/core/cache/backends

**Comment by shaysw** [comment]
> There something weird with your branch - there shouldn't be any conflict with the files: alembic/versions/30862849651c_add_organization_tables.py anyway/error_code_and_strings.py anyway/models.py anyway/scripts/build_redacted_tables_for_dev.py anyway/user_functions.py
> 
> *and maybe some other. as to my understanding this PR shouldn't change them

File: anyway/widgets/suburban_widgets/AccidentCountByDriverTypeWidget.py (hunk 10 above)
Comment (shaysw) [reply anyway/widgets/suburban_widgets/AccidentCountByDriverTypeWidget.py:58]: I totally agree, the thing is - this and other places is existing code, I simply copy-pasted it. I think it is out of scope for this PR to refactor all the code, but I think that it is definitely an effort we should pursue at some point, hopefully sooner than later

File: anyway/widgets/Widget.py (hunk 7 above)
Comment (shaysw) [reply anyway/widgets/Widget.py:90]: I agree, but I didn't want to touch any code, only refactor the structure (see my other reply)

**Comment by shaysw** [comment]
> Regarding naming conventions I think module names are lowercase letters with underscore between words. I think having mixed naming conventions is something we want to avoid.

I agree, and so unless I missed something, my idea was to name modules in lowercase + '_', but name Classes in PascalCasing, which is the convention I think

File: anyway/widgets/suburban_widgets/AccidentCountByAccidentYearWidget.py (hunk 8 above)
Comment (shaysw) [reply anyway/widgets/suburban_widgets/AccidentCountByAccidentYearWidget.py:52]: From what I saw a few times, the exception text gets swallowed somewhere in the process. I can remove it if you think strongly against it, but this is an issue we should probably address at some point 

> Regarding module directory locations - I do not understand the location of `UrbanWidgets.py` and `SubUrbanWidgets.py`. If we adopt a new location standard, we should be applied to other location of base classes.

You mean other classes which are not widgets? I for it, but not in this PR

**Comment by shaysw** [comment]
> There something weird with your branch - there shouldn't be any conflict with the files: alembic/versions/30862849651c_add_organization_tables.py anyway/error_code_and_strings.py anyway/models.py anyway/scripts/build_redacted_tables_for_dev.py anyway/user_functions.py
> 
> *and maybe some other. as to my understanding this PR shouldn't change them

resolved F2F

File: tests/test_infographic_api.py [hunk 11]
```diff
@@ -88,35 +89,38 @@ def test_bad_news_flash_id(self, app):
         assert rv.status_code == http_client.NOT_FOUND
 
     def test_accident_count_by_car_type(self, app):
-        test_involved_by_vehicle_type_data = [{"involve_vehicle_type": 1, "count": 3}, {"involve_vehicle_type": 25, "count": 2},
-         {"involve_vehicle_type": 15, "count": 1}]
+        def mock_get_accidents_stats(table_obj, filters=None, group_by=None, count=None, start_time=None,
```
Comment (shaysw) [inline tests/test_infographic_api.py:92]: try to move out

File: tests/test_infographic_api.py [hunk 12]
```diff
@@ -131,50 +135,25 @@ def mock_get_accidents_stats(table_obj, filters=None, group_by=None, count=None,
...
-                  'percentage_segment': 100.0},
-                 {'car_type': VehicleCategory.MOTORCYCLE.value,
-                  'percentage_country': 6.262912323870099,
-                  'percentage_segment': 0.0},
-                 {'car_type': VehicleCategory.BICYCLE_AND_SMALL_MOTOR.value,
-                  'percentage_country': 2.794865247568421,
-                  'percentage_segment': 0.0}]
-
-        infographics_utils.get_accidents_stats = tmp_func  # Restore function ref - So we don't affect other tests
+                     'percentage_country': 9.84470391606119,
```
Comment (shaysw) [inline tests/test_infographic_api.py:138]: revert formatting changes

**Comment by shaysw** [comment]
> Regarding module directory locations - I do not understand the location of `UrbanWidgets.py` and `SubUrbanWidgets.py`. If we adopt a new location standard, we should be applied to other location of base classes.

Discussed F2F, will continue this structure refactoring to all classes in the repo

**Review by ziv17 — APPROVED** [review]
Nice Work!

**Comment by elazarg** [comment]
This naming scheme is not conventional.

**Pull request merged**